File: axe/parser.py
from .tokens import TokenType
from .errors import ParserError
from .ast import (
    Program,
    ExpressionStatement,
    VariableDeclaration,
    BlockStatement,
    IfStatement,
    ReturnStatement,
    FunctionDeclaration,
    BinaryExpression,
    UnaryExpression,
    Literal,
    Identifier,
    Assignment,
    CallExpression,
)


class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.current = 0

    # ========================================================
    # Entry point
    # ========================================================

    def parse(self):
        statements = []

        self.skip_newlines()

        while not self.is_at_end():
            statements.append(self.statement())
            self.skip_newlines()

        return Program(statements)

    # ========================================================
    # Statements
    # ========================================================

    def statement(self):
        if self.match(TokenType.LET):
            return self.variable_declaration(False)

        if self.match(TokenType.CONST):
            return self.variable_declaration(True)

        if self.match(TokenType.IF):
            return self.if_statement()

        if self.match(TokenType.FN):
            return self.function_declaration()

        if self.match(TokenType.RETURN):
            return self.return_statement()

        return self.expression_statement()

    def variable_declaration(self, constant):
        name = self.consume(
            TokenType.IDENTIFIER,
            "Expected variable name.",
        )

        self.consume(
            TokenType.EQUAL,
            "Expected '=' after variable name.",
        )

        initializer = self.expression()

        return VariableDeclaration(
            name=name.lexeme,
            initializer=initializer,
            constant=constant,
        )

    def expression_statement(self):
        expression = self.expression()

        return ExpressionStatement(expression)

    # ========================================================
    # IF
    # ========================================================

    def if_statement(self):
        condition = self.expression()

        then_branch = self.block()

        else_branch = None

        if self.match(TokenType.ELSE):
            if self.match(TokenType.IF):
                else_branch = self.if_statement()
            else:
                else_branch = self.block()

        return IfStatement(
            condition,
            then_branch,
            else_branch,
        )

    # ========================================================
    # Functions
    # ========================================================

    def function_declaration(self):
        name = self.consume(
            TokenType.IDENTIFIER,
            "Expected function name.",
        )

        self.consume(
            TokenType.LPAREN,
            "Expected '(' after function name.",
        )

        parameters = []

        if not self.check(TokenType.RPAREN):
            while True:
                parameter = self.consume(
                    TokenType.IDENTIFIER,
                    "Expected parameter name.",
                )

                parameters.append(parameter.lexeme)

                if not self.match(TokenType.COMMA):
                    break

        self.consume(
            TokenType.RPAREN,
            "Expected ')' after parameters.",
        )

        body = self.block()

        return FunctionDeclaration(
            name=name.lexeme,
            parameters=parameters,
            body=body,
        )

    # ========================================================
    # Return
    # ========================================================

    def return_statement(self):
        if self.check(TokenType.NEWLINE):
            return ReturnStatement()

        if self.check(TokenType.RBRACE):
            return ReturnStatement()

        return ReturnStatement(
            self.expression()
        )

    # ========================================================
    # Blocks
    # ========================================================

    def block(self):
        self.consume(
            TokenType.LBRACE,
            "Expected '{'.",
        )

        self.skip_newlines()

        statements = []

        while (
            not self.check(TokenType.RBRACE)
            and not self.is_at_end()
        ):
            statements.append(self.statement())
            self.skip_newlines()

        self.consume(
            TokenType.RBRACE,
            "Expected '}'.",
        )

        return BlockStatement(statements)

    # ========================================================
    # Expressions
    # ========================================================

    def expression(self):
        return self.assignment()

    def assignment(self):
        expression = self.logical_or()

        if self.match(TokenType.EQUAL):
            equals = self.previous()
            value = self.assignment()

            if isinstance(expression, Identifier):
                return Assignment(
                    expression.name,
                    value,
                )

            raise self.error(
                equals,
                "Invalid assignment target.",
            )

        return expression

    # ========================================================
    # Logical operators
    # ========================================================

    def logical_or(self):
        expression = self.logical_and()

        while self.match(TokenType.OR):
            operator = self.previous().lexeme
            right = self.logical_and()

            expression = BinaryExpression(
                expression,
                operator,
                right,
            )

        return expression

    def logical_and(self):
        expression = self.equality()

        while self.match(TokenType.AND):
            operator = self.previous().lexeme
            right = self.equality()

            expression = BinaryExpression(
                expression,
                operator,
                right,
            )

        return expression

    # ========================================================
    # Equality
    # ========================================================

    def equality(self):
        expression = self.comparison()

        while self.match(
            TokenType.EQUAL_EQUAL,
            TokenType.NOT_EQUAL,
        ):
            operator = self.previous().lexeme
            right = self.comparison()

            expression = BinaryExpression(
                expression,
                operator,
                right,
            )

        return expression

    # ========================================================
    # Comparison
    # ========================================================

    def comparison(self):
        expression = self.term()

        while self.match(
            TokenType.GREATER,
            TokenType.GREATER_EQUAL,
            TokenType.LESS,
            TokenType.LESS_EQUAL,
        ):
            operator = self.previous().lexeme
            right = self.term()

            expression = BinaryExpression(
                expression,
                operator,
                right,
            )

        return expression

    # ========================================================
    # Addition / subtraction
    # ========================================================

    def term(self):
        expression = self.factor()

        while self.match(
            TokenType.PLUS,
            TokenType.MINUS,
        ):
            operator = self.previous().lexeme
            right = self.factor()

            expression = BinaryExpression(
                expression,
                operator,
                right,
            )

        return expression

    # ========================================================
    # Multiplication
    # ========================================================

    def factor(self):
        expression = self.unary()

        while self.match(
            TokenType.STAR,
            TokenType.SLASH,
            TokenType.PERCENT,
        ):
            operator = self.previous().lexeme
            right = self.unary()

            expression = BinaryExpression(
                expression,
                operator,
                right,
            )

        return expression

    # ========================================================
    # Unary
    # ========================================================

    def unary(self):
        if self.match(
            TokenType.NOT,
            TokenType.MINUS,
        ):
            operator = self.previous().lexeme

            return UnaryExpression(
                operator,
                self.unary(),
            )

        return self.call()

    # ========================================================
    # Function calls
    # ========================================================

    def call(self):
        expression = self.primary()

        while True:
            if self.match(TokenType.LPAREN):
                expression = self.finish_call(expression)
            else:
                break

        return expression

    def finish_call(self, callee):
        arguments = []

        if not self.check(TokenType.RPAREN):
            while True:
                arguments.append(self.expression())

                if not self.match(TokenType.COMMA):
                    break

        self.consume(
            TokenType.RPAREN,
            "Expected ')' after arguments.",
        )

        return CallExpression(
            callee,
            arguments,
        )

    # ========================================================
    # Primary
    # ========================================================

    def primary(self):
        if self.match(TokenType.FALSE):
            return Literal(False)

        if self.match(TokenType.TRUE):
            return Literal(True)

        if self.match(TokenType.NULL):
            return Literal(None)

        if self.match(TokenType.NUMBER):
            return Literal(self.previous().literal)

        if self.match(TokenType.STRING):
            return Literal(self.previous().literal)

        if self.match(TokenType.IDENTIFIER):
            return Identifier(self.previous().lexeme)

        if self.match(TokenType.LPAREN):
            expression = self.expression()

            self.consume(
                TokenType.RPAREN,
                "Expected ')' after expression.",
            )

            return expression

        token = self.peek()

        raise self.error(
            token,
            f"Unexpected token '{token.lexeme}'.",
        )

    # ========================================================
    # Helpers
    # ========================================================

    def match(self, *types):
        for token_type in types:
            if self.check(token_type):
                self.advance()
                return True

        return False

    def consume(self, token_type, message):
        if self.check(token_type):
            return self.advance()

        raise self.error(
            self.peek(),
            message,
        )

    def check(self, token_type):
        if self.is_at_end():
            return token_type == TokenType.EOF

        return self.peek().type == token_type

    def advance(self):
        if not self.is_at_end():
            self.current += 1

        return self.previous()

    def peek(self):
        return self.tokens[self.current]

    def previous(self):
        return self.tokens[self.current - 1]

    def is_at_end(self):
        return self.peek().type == TokenType.EOF

    def skip_newlines(self):
        while self.match(TokenType.NEWLINE):
            pass

    def error(self, token, message):
        return ParserError(
            message,
            token.line,
            token.column,
        )

File: src/parser.rs
use crate::ast::{Expression, LiteralValue, Program, Statement};

use crate::error::NxlError;
use crate::token::{Literal, Token, TokenType};

pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, current: 0 }
    }

    pub fn parse(&mut self) -> Result<Program, NxlError> {
        let mut statements = Vec::new();

        self.skip_newlines();

        while !self.is_at_end() {
            statements.push(self.statement()?);
            self.skip_newlines();
        }

        Ok(Program { statements })
    }

    // ========================================================
    // Statements
    // ========================================================

    fn statement(&mut self) -> Result<Statement, NxlError> {
        if self.match_token(TokenType::Let) {
            return self.variable_declaration(false);
        }

        if self.match_token(TokenType::Const) {
            return self.variable_declaration(true);
        }

        if self.match_token(TokenType::Fn) {
            return self.function_declaration();
        }

        if self.match_token(TokenType::If) {
            return self.if_statement();
        }

        if self.match_token(TokenType::Return) {
            return self.return_statement();
        }

        Ok(Statement::Expression(self.expression()?))
    }

    fn variable_declaration(&mut self, constant: bool) -> Result<Statement, NxlError> {
        let name = self.consume(TokenType::Identifier, "Expected variable name.")?;

        self.consume(TokenType::Equal, "Expected '=' after variable name.")?;

        let initializer = self.expression()?;

        Ok(Statement::VariableDeclaration {
            name: name.lexeme,
            initializer,
            constant,
        })
    }

    // ========================================================
    // Blocks
    // ========================================================

    fn block(&mut self) -> Result<Statement, NxlError> {
        self.consume(TokenType::LeftBrace, "Expected '{'.")?;

        self.skip_newlines();

        let mut statements = Vec::new();

        while !self.check(TokenType::RightBrace) && !self.is_at_end() {
            statements.push(self.statement()?);
            self.skip_newlines();
        }

        self.consume(TokenType::RightBrace, "Expected '}'.")?;

        Ok(Statement::Block { statements })
    }

    // ========================================================
    // If
    // ========================================================

    fn if_statement(&mut self) -> Result<Statement, NxlError> {
        let condition = self.expression()?;

        let then_branch = Box::new(self.block()?);

        let else_branch = if self.match_token(TokenType::Else) {
            Some(Box::new(self.block()?))
        } else {
            None
        };

        Ok(Statement::If {
            condition,
            then_branch,
            else_branch,
        })
    }

    // ========================================================
    // Functions
    // ========================================================

    fn function_declaration(&mut self) -> Result<Statement, NxlError> {
        let name = self.consume(TokenType::Identifier, "Expected function name.")?;

        self.consume(TokenType::LeftParen, "Expected '(' after function name.")?;

        let mut parameters = Vec::new();

        if !self.check(TokenType::RightParen) {
            loop {
                let parameter = self.consume(TokenType::Identifier, "Expected parameter name.")?;

                parameters.push(parameter.lexeme);

                if !self.match_token(TokenType::Comma) {
                    break;
                }
            }
        }

        self.consume(TokenType::RightParen, "Expected ')' after parameters.")?;

        let body = Box::new(self.block()?);

        Ok(Statement::Function {
            name: name.lexeme,
            parameters,
            body,
        })
    }

    // ========================================================
    // Return
    // ========================================================

    fn return_statement(&mut self) -> Result<Statement, NxlError> {
        if self.check(TokenType::Newline) || self.check(TokenType::RightBrace) {
            return Ok(Statement::Return(None));
        }

        let value = self.expression()?;

        Ok(Statement::Return(Some(value)))
    }

    // ========================================================
    // Expressions
    // ========================================================

    fn expression(&mut self) -> Result<Expression, NxlError> {
        self.assignment()
    }

    fn assignment(&mut self) -> Result<Expression, NxlError> {
        let expression = self.logical_or()?;

        if self.match_token(TokenType::Equal) {
            let value = self.assignment()?;

            if let Expression::Identifier(name) = expression {
                return Ok(Expression::Assignment {
                    name,
                    value: Box::new(value),
                });
            }

            return Err(self.error("Invalid assignment target."));
        }

        Ok(expression)
    }

    // ========================================================
    // Logical OR
    // ========================================================

    fn logical_or(&mut self) -> Result<Expression, NxlError> {
        let mut expression = self.logical_and()?;

        while self.match_token(TokenType::Or) {
            let operator = self.previous().token_type.clone();
            let right = self.logical_and()?;

            expression = Expression::Binary {
                left: Box::new(expression),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expression)
    }

    // ========================================================
    // Logical AND
    // ========================================================

    fn logical_and(&mut self) -> Result<Expression, NxlError> {
        let mut expression = self.equality()?;

        while self.match_token(TokenType::And) {
            let operator = self.previous().token_type.clone();
            let right = self.equality()?;

            expression = Expression::Binary {
                left: Box::new(expression),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expression)
    }

    // ========================================================
    // Equality
    // ========================================================

    fn equality(&mut self) -> Result<Expression, NxlError> {
        let mut expression = self.comparison()?;

        while self.match_any(&[TokenType::EqualEqual, TokenType::NotEqual]) {
            let operator = self.previous().token_type.clone();
            let right = self.comparison()?;

            expression = Expression::Binary {
                left: Box::new(expression),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expression)
    }

    // ========================================================
    // Comparison
    // ========================================================

    fn comparison(&mut self) -> Result<Expression, NxlError> {
        let mut expression = self.term()?;

        while self.match_any(&[
            TokenType::Greater,
            TokenType::GreaterEqual,
            TokenType::Less,
            TokenType::LessEqual,
        ]) {
            let operator = self.previous().token_type.clone();
            let right = self.term()?;

            expression = Expression::Binary {
                left: Box::new(expression),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expression)
    }

    // ========================================================
    // Addition / subtraction
    // ========================================================

    fn term(&mut self) -> Result<Expression, NxlError> {
        let mut expression = self.factor()?;

        while self.match_any(&[TokenType::Plus, TokenType::Minus]) {
            let operator = self.previous().token_type.clone();
            let right = self.factor()?;

            expression = Expression::Binary {
                left: Box::new(expression),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expression)
    }

    // ========================================================
    // Multiplication
    // ========================================================

    fn factor(&mut self) -> Result<Expression, NxlError> {
        let mut expression = self.unary()?;

        while self.match_any(&[TokenType::Star, TokenType::Slash, TokenType::Percent]) {
            let operator = self.previous().token_type.clone();
            let right = self.unary()?;

            expression = Expression::Binary {
                left: Box::new(expression),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expression)
    }

    // ========================================================
    // Unary
    // ========================================================

    fn unary(&mut self) -> Result<Expression, NxlError> {
        if self.match_any(&[TokenType::Not, TokenType::Minus]) {
            let operator = self.previous().token_type.clone();
            let operand = self.unary()?;

            return Ok(Expression::Unary {
                operator,
                operand: Box::new(operand),
            });
        }

        self.call()
    }

    // ========================================================
    // Function calls
    // ========================================================

    fn call(&mut self) -> Result<Expression, NxlError> {
        let mut expression = self.primary()?;

        loop {
            if self.match_token(TokenType::LeftParen) {
                expression = self.finish_call(expression)?;
            } else {
                break;
            }
        }

        Ok(expression)
    }

    fn finish_call(&mut self, callee: Expression) -> Result<Expression, NxlError> {
        let mut arguments = Vec::new();

        if !self.check(TokenType::RightParen) {
            loop {
                arguments.push(self.expression()?);

                if !self.match_token(TokenType::Comma) {
                    break;
                }
            }
        }

        self.consume(TokenType::RightParen, "Expected ')' after arguments.")?;

        Ok(Expression::Call {
            callee: Box::new(callee),
            arguments,
        })
    }

    // ========================================================
    // Primary
    // ========================================================

    fn primary(&mut self) -> Result<Expression, NxlError> {
        if self.match_token(TokenType::False) {
            return Ok(Expression::Literal(LiteralValue::Boolean(false)));
        }

        if self.match_token(TokenType::True) {
            return Ok(Expression::Literal(LiteralValue::Boolean(true)));
        }

        if self.match_token(TokenType::Null) {
            return Ok(Expression::Literal(LiteralValue::Null));
        }

        if self.match_token(TokenType::Number) {
            if let Some(Literal::Number(value)) = self.previous().literal.clone() {
                return Ok(Expression::Literal(LiteralValue::Number(value)));
            }
        }

        if self.match_token(TokenType::String) {
            if let Some(Literal::String(value)) = self.previous().literal.clone() {
                return Ok(Expression::Literal(LiteralValue::String(value)));
            }
        }

        if self.match_token(TokenType::Identifier) {
            return Ok(Expression::Identifier(self.previous().lexeme.clone()));
        }

        if self.match_token(TokenType::LeftParen) {
            let expression = self.expression()?;

            self.consume(TokenType::RightParen, "Expected ')' after expression.")?;

            return Ok(expression);
        }

        Err(self.error("Unexpected token."))
    }

    // ========================================================
    // Helpers
    // ========================================================

    fn match_token(&mut self, token_type: TokenType) -> bool {
        if self.check(token_type) {
            self.advance();
            true
        } else {
            false
        }
    }

    fn match_any(&mut self, types: &[TokenType]) -> bool {
        for token_type in types {
            if self.check(token_type.clone()) {
                self.advance();
                return true;
            }
        }

        false
    }

    fn consume(&mut self, token_type: TokenType, message: &str) -> Result<Token, NxlError> {
        if self.check(token_type) {
            return Ok(self.advance());
        }

        Err(self.error(message))
    }

    fn check(&self, token_type: TokenType) -> bool {
        if self.is_at_end() {
            return token_type == TokenType::Eof;
        }

        self.peek().token_type == token_type
    }

    fn advance(&mut self) -> Token {
        if !self.is_at_end() {
            self.current += 1;
        }

        self.previous().clone()
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.current]
    }

    fn previous(&self) -> &Token {
        &self.tokens[self.current - 1]
    }

    fn is_at_end(&self) -> bool {
        self.peek().token_type == TokenType::Eof
    }

    fn skip_newlines(&mut self) {
        while self.match_token(TokenType::Newline) {}
    }

    fn error(&self, message: &str) -> NxlError {
        let token = self.peek();

        NxlError::new(message, token.line, token.column)
    }
}
